src/vs/workbench/api/node/extHostTerminalService.ts

```diff
@@ -212,13 +212,16 @@ export class ExtHostTerminalRenderer extends BaseExtHostTerminal implements vsco
 	constructor(
 		proxy: MainThreadTerminalServiceShape,
 		private _name: string,
-		private _terminal: ExtHostTerminal
+		private _terminal: ExtHostTerminal,
+		id?: number
 	) {
-		super(proxy);
-		this._proxy.$createTerminalRenderer(this._name).then(id => {
-			this._runQueuedRequests(id);
-			(<any>this._terminal)._runQueuedRequests(id);
-		});
+		super(proxy, id);
+		if (!id) {
```

Comment:
@Tyriar I'm not sure about this. Do you have any ideas for how this could be done differently?

Reply:
@GabeDeBacker?

Reply:
The existing terminal implementation did not always create the extension host side of the terminal renderer when a terminal open event was called (through $acceptTerminalOpened). The extension host side renderer must exist and be available before an extension task can begin execution can begin (so it can render output). If you look at the ExtensionCallbackExecutionData.startCallback code (in extHostTasks.ts), it waits for the onDidOpenTerminalRenderer event to actually call into the extension, and it leverages onDidCloseTerminalRenderer as one of the ways to signal cancellation of an extension callback task as well.

In most ways, what this does is actually "complete" the implementation of VSCode's terminal renderer concept as well.

Reply:
It was by design that the `TerminalRenderer` only exists when `createTerminalRenderer` is run. Can the task API just call `createTerminalRenderer` like an extension and derive the `Terminal` from that?

Reply:
Yes, I agree. We should only create the terminal renderer in the extension host if it is needed.

Investigating this.

Reply:
Okay, I've cleaned this up to only create extension host terminal renders when an extension task starts.

We can't use the createTerminalRenderer API as because the task system creates a terminal on a tasks behalf and then makes the extension host aware of it. When the extension callback task starts, it needs to create a renderer to pass to the extension, which is no deferred.

Perhaps creating a terminal renderer for a terminal should be on the terminal itself. One thing that the createTerminalRender and createTerminal APIs don't seem to handle today is the initial dimensions. Is there already an issue for that?